src/gluonnlp/model/transformer.py

```diff
@@ -990,58 +953,180 @@ def decode_seq(self, inputs, states, valid_length=None):
+    output_attention: bool
+        Whether to output the attention weights
+    weight_initializer : str or Initializer
+        Initializer for the input weights matrix, used for the linear
+        transformation of the inputs.
+    bias_initializer : str or Initializer
+        Initializer for the bias vector.
+    scale_embed : bool, default True
+        Scale the input embeddings by sqrt(embed_size).
+    prefix : str, default 'rnn_'
+        Prefix for name of `Block`s
+        (and name of weight if params is `None`).
+    params : Parameter or None
+        Container for weight sharing between cells.
+        Created if `None`.
+    """
+
+
+class TransformerOneStepDecoder(_BaseTransformerDecoder, Seq2SeqOneStepDecoder):
+    """Transformer Decoder.
 
-    def __call__(self, step_input, states): #pylint: disable=arguments-differ
+    One-step ahead decoder for use during inference.
+
+    Parameters
+    ----------
+    attention_cell : AttentionCell or str, default 'multi_head'
+        Arguments of the attention cell.
+        Can be 'multi_head', 'scaled_luong', 'scaled_dot', 'dot', 'cosine', 'normed_mlp', 'mlp'
+    num_layers : int
+        Number of attention layers.
+    units : int
+        Number of units for the output.
+    hidden_size : int
+        number of units in the hidden layer of position-wise feed-forward networks
+    max_length : int
+        Maximum length of the input sequence. This is used for constructing position encoding
+    num_heads : int
+        Number of heads in multi-head attention
+    scaled : bool
+        Whether to scale the softmax input by the sqrt of the input dimension
+        in multi-head attention
+    dropout : float
+        Dropout probability.
+    use_residual : bool
+        Whether to use residual connection.
+    output_attention: bool
+        Whether to output the attention weights
+    weight_initializer : str or Initializer
+        Initializer for the input weights matrix, used for the linear
+        transformation of the inputs.
+    bias_initializer : str or Initializer
+        Initializer for the bias vector.
+    scale_embed : bool, default True
+        Scale the input embeddings by sqrt(embed_size).
+    prefix : str, default 'rnn_'
+        Prefix for name of `Block`s
+        (and name of weight if params is `None`).
+    params : Parameter or None
+        Container for weight sharing between cells.
+        Created if `None`.
+    """
+
+    def forward(self, step_input, states):  # pylint: disable=arguments-differ
+        # We implement forward, as the number of states changes between the
+        # first and later calls of the one-step ahead Transformer decoder. This
+        # is due to the lack of numpy shape semantics. Once we enable numpy
+        # shape semantic in the GluonNLP code-base, the number of states should
+        # stay constant, but the first state element will be an array of shape
+        # (batch_size, 0, C_in) at the first call.
```

Comment:
You mean in the first step, the 'last_embeds' has shape '(batch_size, 0, C_in)'?

Reply:
Yes. Once we enable numpy shape, we can implement it based on `batch_size, 0, C_in` and remove the `def forward` (making it a "real" `HybridBlock`)

Reply:
@leezu Actually in the numpy_nlp implementation I presented today. I used the (B, 0, C_in).

Reply:
I tried 'mx.nd.zeros((3, 0, 2))', but I can get error. How did you successfully create an array with shape (B, 0, C_in)?

Reply:
You have to enable numpy shape first https://github.com/apache/incubator-mxnet/issues/14253 It's opt-in currently (and not supported in GluonNLP)